tensorzero-core/src/optimization/gcp_vertex_gemini_sft/mod.rs

```diff
@@ -0,0 +1,462 @@
+            seed: self.seed,
+            service_account: self.service_account,
+            kms_key_name: self.kms_key_name,
+            tuned_model_display_name: self.tuned_model_display_name,
+            bucket_path_prefix: self.bucket_path_prefix,
+        })
+    }
+}
+
+#[cfg_attr(test, derive(ts_rs::TS))]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(test, ts(export))]
+#[cfg_attr(feature = "pyo3", pyclass(str))]
+pub struct GCPVertexGeminiSFTJobHandle {
+    pub job_id: String,
+    pub job_url: Url,
+    #[cfg_attr(test, ts(type = "string | null"))]
+    pub credential_location: Option<CredentialLocation>,
+    pub region: String,
+    pub project_id: String,
+}
+
+impl std::fmt::Display for GCPVertexGeminiSFTJobHandle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let json = serde_json::to_string_pretty(self).map_err(|_| std::fmt::Error)?;
+        write!(f, "{json}")
+    }
+}
+
+impl Optimizer for GCPVertexGeminiSFTConfig {
+    type Handle = GCPVertexGeminiSFTJobHandle;
+
+    async fn launch(
+        &self,
+        client: &reqwest::Client,
+        train_examples: Vec<RenderedSample>,
+        val_examples: Option<Vec<RenderedSample>>,
+        credentials: &InferenceCredentials,
+    ) -> Result<Self::Handle, Error> {
+        // TODO(#2642): improve error handling here so we know what index of example failed
+        let train_rows: Vec<GCPVertexGeminiSupervisedRow> = train_examples
+            .iter()
+            .map(GCPVertexGeminiSupervisedRow::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let val_rows: Option<Vec<GCPVertexGeminiSupervisedRow>> = val_examples
+            .as_ref()
+            .map(|examples| {
+                examples
+                    .iter()
+                    .map(GCPVertexGeminiSupervisedRow::try_from)
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .transpose()?;
+
+        let train_filename = format!("train_{}.jsonl", uuid::Uuid::now_v7()); // or use job ID
+        let train_gs_url = match &self.bucket_path_prefix {
+            Some(prefix) => format!("gs://{}/{}/{}", self.bucket_name, prefix, train_filename),
+            None => format!("gs://{}/{}", self.bucket_name, train_filename),
+        };
+        upload_rows_to_gcp_object_store(&train_rows, &train_gs_url, &self.credentials, credentials)
+            .await?;
+
+        // Upload validation data if provided
+        let val_gs_url = if let Some(val_rows) = &val_rows {
+            let val_filename = format!("val_{}.jsonl", uuid::Uuid::now_v7());
+            let val_url = match &self.bucket_path_prefix {
+                Some(prefix) => format!("gs://{}/{}/{}", self.bucket_name, prefix, val_filename),
+                None => format!("gs://{}/{}", self.bucket_name, val_filename),
+            };
+
+            upload_rows_to_gcp_object_store(val_rows, &val_url, &self.credentials, credentials)
```

Comment:
This upload can be done in parallel with the training file by avoiding using `.await` directly on the function result, and using `try_join`. See `FireworksSFTConfig.launch` for an example